packages/notebook/src/history.ts

```diff
@@ -0,0 +1,420 @@
+    this._isDisposed = true;
+    this._history.length = 0;
+    Signal.clearData(this);
+  }
+
+  /**
+   * Set placeholder and editor. Start session if one is not already started.
+   *
+   * @param activeCell - The currently selected Cell in the notebook.
+   */
+  protected async checkSession(activeCell: Cell): Promise<void> {
+    if (!this._hasSession) {
+      await this._retrieveHistory();
+      this._hasSession = true;
+      this.editor = activeCell.editor;
+      this._placeholder = this._editor?.model.sharedModel.getSource() || '';
+      // Filter the history with the placeholder string.
+      this.setFilter(this._placeholder);
+      this._cursor = this._filtered.length - 1;
+    }
+  }
+
+  /**
+   * Get the previous item in the console history.
+   *
+   * @param activeCell - The currently selected Cell in the notebook.
+   *
+   * @returns A Promise for console command text or `undefined` if unavailable.
+   */
+  async back(activeCell: Cell): Promise<string> {
+    await this.checkSession(activeCell);
+    --this._cursor;
+    if (this._cursor < 0) {
+      await this.fetchBatch();
+    }
+    this._cursor = Math.max(0, this._cursor);
+    const content = this._filtered[this._cursor];
+    return content;
+  }
+
+  /**
+   * Get the next item in the console history.
+   *
+   * @param activeCell - The currently selected Cell in the notebook.
+   *
+   * @returns A Promise for console command text or `undefined` if unavailable.
+   */
+  async forward(activeCell: Cell): Promise<string> {
+    await this.checkSession(activeCell);
+    ++this._cursor;
+    this._cursor = Math.min(this._filtered.length - 1, this._cursor);
+    const content = this._filtered[this._cursor];
+    return content;
+  }
+
+  /**
+   * Get the next item in the console history.
+   *
+   * @param activeCell - The currently selected Cell in the notebook.
+   * @param update - the result from back or forward
+   */
+  updateEditor(activeCell: Cell, update: string): void {
+    if (activeCell) {
+      const model = activeCell.editor?.model;
+      const source = model?.sharedModel.getSource();
+      if (this.isDisposed || !update) {
+        return;
+      }
+      if (source === update) {
+        return;
+      }
+      this._setByHistory = true;
+      model?.sharedModel.setSource(update);
+      let columnPos = 0;
+      columnPos = update.indexOf('\n');
+      if (columnPos < 0) {
+        columnPos = update.length;
+      }
+      activeCell.editor?.setCursorPosition({ line: 0, column: columnPos });
+    }
+  }
+
+  /**
+   * Reset the history navigation state, i.e., start a new history session.
+   */
+  reset(): void {
+    this._hasSession = false;
+    this._placeholder = '';
+    this.toRequest = this.requestBatchSize;
+  }
+
+  /**
+   * Fetches a subsequent batch of history. Updates the filtered history and cursor to correct place in history,
+   * accounting for potentially new history items above it.
+   */
+  async fetchBatch() {
```

Comment:
private/protected?

Would it be worth adding a jest unit test for this logic (indirectly, by testing `back` which is the public API for it)?

Reply:
Good call. I changed it to private, and added a test as well